_published/0013-authentication.md

```diff
@@ -0,0 +1,165 @@
+possible to either include the ID token in every request to the server or to
+start using sessions to store the ID token between requests.
+
+#### Local development vs production environments
+For the initial release of the feature the configuration will be tailored to
+using azure as an IDP. These settings will as necessary be exposed as
+environment variables to each respective docker image.
+
+
+#### Choice of Client
+
+The OIDC standard defines a
+[list of certified clients](https://openid.net/developers/certified/)
+that can help simplify the implementation of ID Token retrieval.
+
+For the backend a full client will not be required. As the tokens just need
+to be validated and not requested. Validation can be done offline by caching
+the issuers' signature for future validations (OBS! The signature key may
+rotate on a timely basis).
+
+## Compatibility
+
+Expect no compatability between versions of wharf without authentication to
+versions of wharf with authentication. There is a potential threat in leaving
+the system backwards compatible in that a non-secured version of the api could
+run together in an otherwise secure cluster.
+
+We should aim to pick a solution where any IDP supporting OIDC and RBAC can be
+used. Locking ourselves to a single IDP when a standard such as OIDC exists
+seems like a  waste opportunity. Keeping the choice of IDP open will leave
+more room for choice and flexibility in the future.
+
+## Alternative solutions
+
+There are many types of identity verification systems.
+This RFC assumes that OIDC will be used. Within OIDC there are multiple
+different [flow definitions](https://developer.okta.com/docs/concepts/oauth-openid/#choosing-an-oauth-2-0-flow).
+We should primarily use `codeflow with pkce` as this is presumed to be the
+most secure, and best compatible at the time of writing.
+
+## Future possibilities
+
+We may want to implement multiple different IDP sources. This could allow
+login from different sources than the Iver-Azure-AD. A sample situation
+where this could be useful is where/when we wish to give a client
+view/control over the deployment pipeline.
+
+When handling ID-Tokens from RBAC enabled IDPs its is possible to verify
+custom permission levels using custom role claims.
+[See this link](https://docs.microsoft.com/en-us/azure/role-based-access-control/custom-roles)
+for more information about custom claims. This system allows for setup
+fine-tuned access control on a per-deployment basis. The initial setup
+for each of the different RBACs will likely require setup scripts on a
+per-provider basis as it requires interaction directly with each respective
+cloud on a non-standard basis.
+
+## Unresolved questions
+
+OIDC in any form demands full SSL security in all connections. In order to
+avoid having to fill the development environment with --force-unsecure
+bindings we should consider implementing certificate generation and importing
+for dev environments.
```

Comment:
We can tell traefik to use a locally generated certificate. Ex: https://github.com/traefik/traefik/issues/5507#issuecomment-535603881. Need to be careful not to use the same cert across all us devs by storing it in the repo, as that would be an insane attack vector.

But instead we could have an extra step to the docker-compose setup, something like:

1. Generate cert using openssl, for example as defined in a `Makefile`
2. Trust the newly generated certificate, with different docs for windows and linux. For example in linux it would be something like this (which could also be inserted into a `Makefile`):
   ```sh
   sudo cp wharf-local-dev.crt /usr/local/share/ca-certificates/wharf-local-dev.crt
   sudo update-ca-certificates
   ```
3. ```sh
   docker-compose up
   ```
Voila

Reply:
Good plan! Will write this into a todo for the wharf-docker-compose repo as a subtask of auth.

Reply:
Perhaps the domain name `wharf.local` might be a good starting point aswell. Probably an easy switch later in implementation.

Reply:
Potentially add a frontend warning if communicating on http non-secure

Reply:
Given localhost is possible that seems much better.

Reply:
This is semi rewritten for the subtitle  `### SSL Config Required as a part of OIDC` now.